# Prosper Loan Data Project 

## Structure of dataset

#### The Prosper loan dataset comprises of 113937 loan entries with 81 attributes on each loan, including loan amount, borrower rate (or interest rate), current loan status, borrower income, and many others, from the year 2009-2014. There are two main categories:

* Borrower information: Basic attributes of the borrowers such as annual income, condition of employment, interest rate, loan status, etc.
* Loan performance information: Metrics evaluating the risk associated with the loans such as Prosper score and bank card utilization, etc.

In [1]:
#Basic libraries for data wranggling

import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
import pandas as pd

#Library for ML model
import sklearn

# Ignore warning messages
import warnings
warnings.filterwarnings('ignore')
#step

In [2]:
#lOADING OF THE DATASET
df=pd.read_csv("prosperLoanData.csv")
df.head()

ListingKey  ListingNumber            ListingCreationDate  \
0  1021339766868145413AB3B         193129  2007-08-26 19:09:29.263000000   
1  10273602499503308B223C1        1209647  2014-02-27 08:28:07.900000000   
2  0EE9337825851032864889A          81716  2007-01-05 15:00:47.090000000   
3  0EF5356002482715299901A         658116  2012-10-22 11:02:35.010000000   
4  0F023589499656230C5E3E2         909464  2013-09-14 18:38:39.097000000   

  CreditGrade  Term LoanStatus           ClosedDate  BorrowerAPR  \
0           C    36  Completed  2009-08-14 00:00:00      0.16516   
1         NaN    36    Current                  NaN      0.12016   
2          HR    36  Completed  2009-12-17 00:00:00      0.28269   
3         NaN    36    Current                  NaN      0.12528   
4         NaN    36    Current                  NaN      0.24614   

   BorrowerRate  LenderYield  ...  LP_ServiceFees  LP_CollectionFees  \
0        0.1580       0.1380  ...         -133.18                0.0   
1        0.0920       0.0820  ...            0.00                0.0   
2        0.2750       0.2400  ...          -24.20                0.0   
3        0.0974       0.0874  ...         -108.01                0.0   
4        0.2085       0.1985  ...          -60.27                0.0   

   LP_GrossPrincipalLoss  LP_NetPrincipalLoss LP_NonPrincipalRecoverypayments  \
0                    0.0                  0.0                             0.0   
1                    0.0                  0.0                             0.0   
2                    0.0                  0.0                             0.0   
3                    0.0                  0.0                             0.0   
4                    0.0                  0.0                             0.0   

   PercentFunded  Recommendations InvestmentFromFriendsCount  \
0            1.0                0                          0   
1            1.0                0                          0   
2            1.0                0                          0   
3            1.0                0                          0   
4            1.0                0                          0   

  InvestmentFromFriendsAmount Investors  
0                         0.0       258  
1                         0.0         1  
2                         0.0        41  
3                         0.0       158  
4                         0.0        20  

[5 rows x 81 columns]

In [3]:
df.shape

(113937, 81)

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 113937 entries, 0 to 113936
Data columns (total 81 columns):
 #   Column                               Non-Null Count   Dtype  
---  ------                               --------------   -----  
 0   ListingKey                           113937 non-null  object 
 1   ListingNumber                        113937 non-null  int64  
 2   ListingCreationDate                  113937 non-null  object 
 3   CreditGrade                          28953 non-null   object 
 4   Term                                 113937 non-null  int64  
 5   LoanStatus                           113937 non-null  object 
 6   ClosedDate                           55089 non-null   object 
 7   BorrowerAPR                          113912 non-null  float64
 8   BorrowerRate                         113937 non-null  float64
 9   LenderYield                          113937 non-null  float64
 10  EstimatedEffectiveYield              84853 non-null   float64
 11  EstimatedLoss

In [5]:
df.describe()

ListingNumber           Term    BorrowerAPR   BorrowerRate  \
count   1.139370e+05  113937.000000  113912.000000  113937.000000   
mean    6.278857e+05      40.830248       0.218828       0.192764   
std     3.280762e+05      10.436212       0.080364       0.074818   
min     4.000000e+00      12.000000       0.006530       0.000000   
25%     4.009190e+05      36.000000       0.156290       0.134000   
50%     6.005540e+05      36.000000       0.209760       0.184000   
75%     8.926340e+05      36.000000       0.283810       0.250000   
max     1.255725e+06      60.000000       0.512290       0.497500   

         LenderYield  EstimatedEffectiveYield  EstimatedLoss  EstimatedReturn  \
count  113937.000000             84853.000000   84853.000000     84853.000000   
mean        0.182701                 0.168661       0.080306         0.096068   
std         0.074516                 0.068467       0.046764         0.030403   
min        -0.010000                -0.182700       0.004900        -0.182700   
25%         0.124200                 0.115670       0.042400         0.074080   
50%         0.173000                 0.161500       0.072400         0.091700   
75%         0.240000                 0.224300       0.112000         0.116600   
max         0.492500                 0.319900       0.366000         0.283700   

       ProsperRating (numeric)  ProsperScore  ...  LP_ServiceFees  \
count             84853.000000  84853.000000  ...   113937.000000   
mean                  4.072243      5.950067  ...      -54.725641   
std                   1.673227      2.376501  ...       60.675425   
min                   1.000000      1.000000  ...     -664.870000   
25%                   3.000000      4.000000  ...      -73.180000   
50%                   4.000000      6.000000  ...      -34.440000   
75%                   5.000000      8.000000  ...      -13.920000   
max                   7.000000     11.000000  ...       32.060000   

       LP_CollectionFees  LP_GrossPrincipalLoss  LP_NetPrincipalLoss  \
count      113937.000000          113937.000000        113937.000000   
mean          -14.242698             700.446342           681.420499   
std           109.232758            2388.513831          2357.167068   
min         -9274.750000             -94.200000          -954.550000   
25%             0.000000               0.000000             0.000000   
50%             0.000000               0.000000             0.000000   
75%             0.000000               0.000000             0.000000   
max             0.000000           25000.000000         25000.000000   

       LP_NonPrincipalRecoverypayments  PercentFunded  Recommendations  \
count                    113937.000000  113937.000000    113937.000000   
mean                         25.142686       0.998584         0.048027   
std                         275.657937       0.017919         0.332353   
min                           0.000000       0.700000         0.000000   
25%                           0.000000       1.000000         0.000000   
50%                           0.000000       1.000000         0.000000   
75%                           0.000000       1.000000         0.000000   
max                       21117.900000       1.012500        39.000000   

       InvestmentFromFriendsCount  InvestmentFromFriendsAmount      Investors  
count               113937.000000                113937.000000  113937.000000  
mean                     0.023460                    16.550751      80.475228  
std                      0.232412                   294.545422     103.239020  
min                      0.000000                     0.000000       1.000000  
25%                      0.000000                     0.000000       2.000000  
50%                      0.000000                     0.000000      44.000000  
75%                      0.000000                     0.000000     115.000000  
max                     33.000000                 25000.000000    1189.000000

## Preprocessing of the Data

Some columns have to processed as they contain dirty data 

In [6]:
df.isna().sum()
#checking of the null values 

ListingKey                         0
ListingNumber                      0
ListingCreationDate                0
CreditGrade                    84984
Term                               0
                               ...  
PercentFunded                      0
Recommendations                    0
InvestmentFromFriendsCount         0
InvestmentFromFriendsAmount        0
Investors                          0
Length: 81, dtype: int64

In [7]:
#some columns are seleccted to make a subdataset and for cleaning of data 

col_feature=['ListingNumber','ListingCreationDate','ProsperRating (Alpha)','ProsperScore','ListingCategory (numeric)', 'Occupation','EmploymentStatus', 
            'IsBorrowerHomeowner', 'CurrentCreditLines', 'OpenRevolvingAccounts', 'DebtToIncomeRatio','IncomeVerifiable', 
            'LoanOriginalAmount', 'LoanStatus','CreditScoreRangeLower', 'CreditScoreRangeUpper','BorrowerState', 
            'LoanOriginationDate', 'IncomeRange','BorrowerAPR','BorrowerRate','StatedMonthlyIncome','Term']

In [8]:
#new dataset with limited features extracted from the old dataset
loan_df=df[col_feature]
loan_df.head()

ListingNumber            ListingCreationDate ProsperRating (Alpha)  \
0         193129  2007-08-26 19:09:29.263000000                   NaN   
1        1209647  2014-02-27 08:28:07.900000000                     A   
2          81716  2007-01-05 15:00:47.090000000                   NaN   
3         658116  2012-10-22 11:02:35.010000000                     A   
4         909464  2013-09-14 18:38:39.097000000                     D   

   ProsperScore  ListingCategory (numeric)     Occupation EmploymentStatus  \
0           NaN                          0          Other    Self-employed   
1           7.0                          2   Professional         Employed   
2           NaN                          0          Other    Not available   
3           9.0                         16  Skilled Labor         Employed   
4           4.0                          2      Executive         Employed   

   IsBorrowerHomeowner  CurrentCreditLines  OpenRevolvingAccounts  ...  \
0                 True                 5.0                      1  ...   
1                False                14.0                     13  ...   
2                False                 NaN                      0  ...   
3                 True                 5.0                      7  ...   
4                 True                19.0                      6  ...   

   LoanStatus  CreditScoreRangeLower  CreditScoreRangeUpper BorrowerState  \
0   Completed                  640.0                  659.0            CO   
1     Current                  680.0                  699.0            CO   
2   Completed                  480.0                  499.0            GA   
3     Current                  800.0                  819.0            GA   
4     Current                  680.0                  699.0            MN   

   LoanOriginationDate     IncomeRange BorrowerAPR BorrowerRate  \
0  2007-09-12 00:00:00  $25,000-49,999     0.16516       0.1580   
1  2014-03-03 00:00:00  $50,000-74,999     0.12016       0.0920   
2  2007-01-17 00:00:00   Not displayed     0.28269       0.2750   
3  2012-11-01 00:00:00  $25,000-49,999     0.12528       0.0974   
4  2013-09-20 00:00:00       $100,000+     0.24614       0.2085   

  StatedMonthlyIncome  Term  
0         3083.333333    36  
1         6125.000000    36  
2         2083.333333    36  
3         2875.000000    36  
4         9583.333333    36  

[5 rows x 23 columns]

In [9]:
loan_df.shape

(113937, 23)

In [10]:
# info of the current dataset
loan_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 113937 entries, 0 to 113936
Data columns (total 23 columns):
 #   Column                     Non-Null Count   Dtype  
---  ------                     --------------   -----  
 0   ListingNumber              113937 non-null  int64  
 1   ListingCreationDate        113937 non-null  object 
 2   ProsperRating (Alpha)      84853 non-null   object 
 3   ProsperScore               84853 non-null   float64
 4   ListingCategory (numeric)  113937 non-null  int64  
 5   Occupation                 110349 non-null  object 
 6   EmploymentStatus           111682 non-null  object 
 7   IsBorrowerHomeowner        113937 non-null  bool   
 8   CurrentCreditLines         106333 non-null  float64
 9   OpenRevolvingAccounts      113937 non-null  int64  
 10  DebtToIncomeRatio          105383 non-null  float64
 11  IncomeVerifiable           113937 non-null  bool   
 12  LoanOriginalAmount         113937 non-null  int64  
 13  LoanStatus                 11

In [11]:
loan_df.describe()

ListingNumber  ProsperScore  ListingCategory (numeric)  \
count   1.139370e+05  84853.000000              113937.000000   
mean    6.278857e+05      5.950067                   2.774209   
std     3.280762e+05      2.376501                   3.996797   
min     4.000000e+00      1.000000                   0.000000   
25%     4.009190e+05      4.000000                   1.000000   
50%     6.005540e+05      6.000000                   1.000000   
75%     8.926340e+05      8.000000                   3.000000   
max     1.255725e+06     11.000000                  20.000000   

       CurrentCreditLines  OpenRevolvingAccounts  DebtToIncomeRatio  \
count       106333.000000           113937.00000      105383.000000   
mean            10.317192                6.96979           0.275947   
std              5.457866                4.63097           0.551759   
min              0.000000                0.00000           0.000000   
25%              7.000000                4.00000           0.140000   
50%             10.000000                6.00000           0.220000   
75%             13.000000                9.00000           0.320000   
max             59.000000               51.00000          10.010000   

       LoanOriginalAmount  CreditScoreRangeLower  CreditScoreRangeUpper  \
count        113937.00000          113346.000000          113346.000000   
mean           8337.01385             685.567731             704.567731   
std            6245.80058              66.458275              66.458275   
min            1000.00000               0.000000              19.000000   
25%            4000.00000             660.000000             679.000000   
50%            6500.00000             680.000000             699.000000   
75%           12000.00000             720.000000             739.000000   
max           35000.00000             880.000000             899.000000   

         BorrowerAPR   BorrowerRate  StatedMonthlyIncome           Term  
count  113912.000000  113937.000000         1.139370e+05  113937.000000  
mean        0.218828       0.192764         5.608026e+03      40.830248  
std         0.080364       0.074818         7.478497e+03      10.436212  
min         0.006530       0.000000         0.000000e+00      12.000000  
25%         0.156290       0.134000         3.200333e+03      36.000000  
50%         0.209760       0.184000         4.666667e+03      36.000000  
75%         0.283810       0.250000         6.825000e+03      36.000000  
max         0.512290       0.497500         1.750003e+06      60.000000

In [12]:
# checking out the null values 
loan_df.isna().sum()

ListingNumber                    0
ListingCreationDate              0
ProsperRating (Alpha)        29084
ProsperScore                 29084
ListingCategory (numeric)        0
Occupation                    3588
EmploymentStatus              2255
IsBorrowerHomeowner              0
CurrentCreditLines            7604
OpenRevolvingAccounts            0
DebtToIncomeRatio             8554
IncomeVerifiable                 0
LoanOriginalAmount               0
LoanStatus                       0
CreditScoreRangeLower          591
CreditScoreRangeUpper          591
BorrowerState                 5515
LoanOriginationDate              0
IncomeRange                      0
BorrowerAPR                     25
BorrowerRate                     0
StatedMonthlyIncome              0
Term                             0
dtype: int64

##### so as I can see from the above code that major null values are present in "ProsperRating (Alpha)" i.e. 29084 and ProsperScore also same they have to removed 

In [13]:
# Check for duplicates in the data
sum(loan_df.duplicated())

0

In [14]:
# Returns the number of unique in the dataset
loan_df.nunique()

ListingNumber                113066
ListingCreationDate          113064
ProsperRating (Alpha)             7
ProsperScore                     11
ListingCategory (numeric)        21
Occupation                       67
EmploymentStatus                  8
IsBorrowerHomeowner               2
CurrentCreditLines               54
OpenRevolvingAccounts            48
DebtToIncomeRatio              1207
IncomeVerifiable                  2
LoanOriginalAmount             2468
LoanStatus                       12
CreditScoreRangeLower            26
CreditScoreRangeUpper            26
BorrowerState                    51
LoanOriginationDate            1873
IncomeRange                       8
BorrowerAPR                    6677
BorrowerRate                   2294
StatedMonthlyIncome           13502
Term                              3
dtype: int64

In [15]:
# According to my analysis, ProsperRating is the most important feature of this dataset so i will select only not null rows

loan_df=loan_df[loan_df['ProsperRating (Alpha)'].notnull()]

In [16]:
loan_df.shape

(84853, 23)

In [17]:
# Filling missing values with mean
loan_df['DebtToIncomeRatio']=loan_df['DebtToIncomeRatio'].fillna(loan_df['DebtToIncomeRatio'].mean())

In [18]:
# Filling missing values 
loan_df['Occupation']=loan_df['Occupation'].fillna('Unknown')

In [19]:
loan_df.nunique()

ListingNumber                83982
ListingCreationDate          83980
ProsperRating (Alpha)            7
ProsperScore                    11
ListingCategory (numeric)       20
Occupation                      68
EmploymentStatus                 7
IsBorrowerHomeowner              2
CurrentCreditLines              51
OpenRevolvingAccounts           46
DebtToIncomeRatio              260
IncomeVerifiable                 2
LoanOriginalAmount            1934
LoanStatus                      11
CreditScoreRangeLower           15
CreditScoreRangeUpper           15
BorrowerState                   48
LoanOriginationDate           1165
IncomeRange                      7
BorrowerAPR                   2330
BorrowerRate                  1229
StatedMonthlyIncome           9445
Term                             3
dtype: int64

In [20]:
loan_df.isna().sum()

ListingNumber                0
ListingCreationDate          0
ProsperRating (Alpha)        0
ProsperScore                 0
ListingCategory (numeric)    0
Occupation                   0
EmploymentStatus             0
IsBorrowerHomeowner          0
CurrentCreditLines           0
OpenRevolvingAccounts        0
DebtToIncomeRatio            0
IncomeVerifiable             0
LoanOriginalAmount           0
LoanStatus                   0
CreditScoreRangeLower        0
CreditScoreRangeUpper        0
BorrowerState                0
LoanOriginationDate          0
IncomeRange                  0
BorrowerAPR                  0
BorrowerRate                 0
StatedMonthlyIncome          0
Term                         0
dtype: int64

In [21]:
loan_df.head()

ListingNumber            ListingCreationDate ProsperRating (Alpha)  \
1        1209647  2014-02-27 08:28:07.900000000                     A   
3         658116  2012-10-22 11:02:35.010000000                     A   
4         909464  2013-09-14 18:38:39.097000000                     D   
5        1074836  2013-12-14 08:26:37.093000000                     B   
6         750899  2013-04-12 09:52:56.147000000                     E   

   ProsperScore  ListingCategory (numeric)      Occupation EmploymentStatus  \
1           7.0                          2    Professional         Employed   
3           9.0                         16   Skilled Labor         Employed   
4           4.0                          2       Executive         Employed   
5          10.0                          1    Professional         Employed   
6           2.0                          1  Sales - Retail         Employed   

   IsBorrowerHomeowner  CurrentCreditLines  OpenRevolvingAccounts  ...  \
1                False                14.0                     13  ...   
3                 True                 5.0                      7  ...   
4                 True                19.0                      6  ...   
5                 True                21.0                     13  ...   
6                False                10.0                      6  ...   

   LoanStatus  CreditScoreRangeLower  CreditScoreRangeUpper BorrowerState  \
1     Current                  680.0                  699.0            CO   
3     Current                  800.0                  819.0            GA   
4     Current                  680.0                  699.0            MN   
5     Current                  740.0                  759.0            NM   
6     Current                  680.0                  699.0            KS   

   LoanOriginationDate     IncomeRange BorrowerAPR BorrowerRate  \
1  2014-03-03 00:00:00  $50,000-74,999     0.12016       0.0920   
3  2012-11-01 00:00:00  $25,000-49,999     0.12528       0.0974   
4  2013-09-20 00:00:00       $100,000+     0.24614       0.2085   
5  2013-12-24 00:00:00       $100,000+     0.15425       0.1314   
6  2013-04-18 00:00:00  $25,000-49,999     0.31032       0.2712   

  StatedMonthlyIncome  Term  
1         6125.000000    36  
3         2875.000000    36  
4         9583.333333    36  
5         8333.333333    60  
6         2083.333333    36  

[5 rows x 23 columns]

In [22]:
# check unique loan status 
loan_df["LoanStatus"].unique()

array(['Current', 'Past Due (1-15 days)', 'Defaulted', 'Completed',
       'Chargedoff', 'Past Due (16-30 days)', 'Past Due (61-90 days)',
       'Past Due (31-60 days)', 'Past Due (91-120 days)',
       'FinalPaymentInProgress', 'Past Due (>120 days)'], dtype=object)

In [23]:
#checking the dtypes of every column
loan_df.dtypes

ListingNumber                  int64
ListingCreationDate           object
ProsperRating (Alpha)         object
ProsperScore                 float64
ListingCategory (numeric)      int64
Occupation                    object
EmploymentStatus              object
IsBorrowerHomeowner             bool
CurrentCreditLines           float64
OpenRevolvingAccounts          int64
DebtToIncomeRatio            float64
IncomeVerifiable                bool
LoanOriginalAmount             int64
LoanStatus                    object
CreditScoreRangeLower        float64
CreditScoreRangeUpper        float64
BorrowerState                 object
LoanOriginationDate           object
IncomeRange                   object
BorrowerAPR                  float64
BorrowerRate                 float64
StatedMonthlyIncome          float64
Term                           int64
dtype: object

In [24]:
# Convert 'ListingCreationDate' type from str to datetime 
loan_df['ListingCreationDate']=pd.to_datetime(loan_df['ListingCreationDate'])

In [25]:
loan_df["ListingCreationDate"]


1        2014-02-27 08:28:07.900
3        2012-10-22 11:02:35.010
4        2013-09-14 18:38:39.097
5        2013-12-14 08:26:37.093
6        2013-04-12 09:52:56.147
                   ...          
113932   2013-04-14 05:55:02.663
113933   2011-11-03 20:42:55.333
113934   2013-12-13 05:49:12.703
113935   2011-11-14 13:18:26.597
113936   2014-01-15 09:27:37.657
Name: ListingCreationDate, Length: 84853, dtype: datetime64[ns]

In [26]:
loan_df["LoanOriginationDate"]

1         2014-03-03 00:00:00
3         2012-11-01 00:00:00
4         2013-09-20 00:00:00
5         2013-12-24 00:00:00
6         2013-04-18 00:00:00
                 ...         
113932    2013-04-22 00:00:00
113933    2011-11-07 00:00:00
113934    2013-12-23 00:00:00
113935    2011-11-21 00:00:00
113936    2014-01-21 00:00:00
Name: LoanOriginationDate, Length: 84853, dtype: object

In [27]:
# Extract date information from ListingCreationDate column
loan_df['date'] = loan_df['ListingCreationDate'].dt.date

# Extract year information from ListingCreationDate column
loan_df['year']= loan_df['ListingCreationDate'].dt.year

# Extract month information from ListingCreationDate column
loan_df['month'] = loan_df['ListingCreationDate'].dt.month

# Convert 'date' type from str to datetime 
loan_df['date']=pd.to_datetime(loan_df['date'])

# Drop unused column
loan_df.drop(['ListingCreationDate'], axis=1, inplace=True)

In [28]:
# unique values in the month column

loan_df['month'].unique()

array([ 2, 10,  9, 12,  4,  5,  7,  8, 11,  1,  6,  3], dtype=int64)

In [29]:
loan_df.date.unique()

array(['2014-02-27T00:00:00.000000000', '2012-10-22T00:00:00.000000000',
       '2013-09-14T00:00:00.000000000', ...,
       '2009-12-26T00:00:00.000000000', '2010-12-25T00:00:00.000000000',
       '2010-09-05T00:00:00.000000000'], dtype='datetime64[ns]')

In [30]:
loan_df.year.unique()

array([2014, 2012, 2013, 2010, 2011, 2009], dtype=int64)

In [31]:
# Replacing numerical values of month with names
# Convert 'Month' type  to str
loan_df['month']=loan_df['month'].astype(str)

# replacing numerical values of month with month names
loan_df.month=loan_df.month.replace([ '1', '2','3','4','5','6','7','8', '9','10','11','12'], ['Jan','Feb','Mar','Apr','May','Jun','Jul','Aug','Sept','Oct','Nov','Dec'])

# check the correction
loan_df['month'].unique()

array(['Feb', 'Oct', 'Sept', 'Dec', 'Apr', 'May', 'Jul', 'Aug', 'Nov',
       'Jan', 'Jun', 'Mar'], dtype=object)

In [32]:
loan_df.head()

ListingNumber ProsperRating (Alpha)  ProsperScore  \
1        1209647                     A           7.0   
3         658116                     A           9.0   
4         909464                     D           4.0   
5        1074836                     B          10.0   
6         750899                     E           2.0   

   ListingCategory (numeric)      Occupation EmploymentStatus  \
1                          2    Professional         Employed   
3                         16   Skilled Labor         Employed   
4                          2       Executive         Employed   
5                          1    Professional         Employed   
6                          1  Sales - Retail         Employed   

   IsBorrowerHomeowner  CurrentCreditLines  OpenRevolvingAccounts  \
1                False                14.0                     13   
3                 True                 5.0                      7   
4                 True                19.0                      6   
5                 True                21.0                     13   
6                False                10.0                      6   

   DebtToIncomeRatio  ...  BorrowerState  LoanOriginationDate     IncomeRange  \
1               0.18  ...             CO  2014-03-03 00:00:00  $50,000-74,999   
3               0.15  ...             GA  2012-11-01 00:00:00  $25,000-49,999   
4               0.26  ...             MN  2013-09-20 00:00:00       $100,000+   
5               0.36  ...             NM  2013-12-24 00:00:00       $100,000+   
6               0.27  ...             KS  2013-04-18 00:00:00  $25,000-49,999   

   BorrowerAPR  BorrowerRate StatedMonthlyIncome Term       date  year  month  
1      0.12016        0.0920         6125.000000   36 2014-02-27  2014    Feb  
3      0.12528        0.0974         2875.000000   36 2012-10-22  2012    Oct  
4      0.24614        0.2085         9583.333333   36 2013-09-14  2013   Sept  
5      0.15425        0.1314         8333.333333   60 2013-12-14  2013    Dec  
6      0.31032        0.2712         2083.333333   36 2013-04-12  2013    Apr  

[5 rows x 25 columns]

In [33]:
# Convert 'LoanOriginationDate' type from str to datetime 
loan_df['LoanOriginationDate']=pd.to_datetime(loan_df['LoanOriginationDate'])

In [34]:
loan_df['ListingNumber']=loan_df['ListingNumber'].astype(str)
loan_df['ProsperScore']=loan_df['ProsperScore'].astype(int)

In [35]:
loan_df.dtypes

ListingNumber                        object
ProsperRating (Alpha)                object
ProsperScore                          int32
ListingCategory (numeric)             int64
Occupation                           object
EmploymentStatus                     object
IsBorrowerHomeowner                    bool
CurrentCreditLines                  float64
OpenRevolvingAccounts                 int64
DebtToIncomeRatio                   float64
IncomeVerifiable                       bool
LoanOriginalAmount                    int64
LoanStatus                           object
CreditScoreRangeLower               float64
CreditScoreRangeUpper               float64
BorrowerState                        object
LoanOriginationDate          datetime64[ns]
IncomeRange                          object
BorrowerAPR                         float64
BorrowerRate                        float64
StatedMonthlyIncome                 float64
Term                                  int64
date                         dat

In [36]:
loan_df["ListingCategory (numeric)"]

1          2
3         16
4          2
5          1
6          1
          ..
113932     1
113933     7
113934     1
113935     2
113936     1
Name: ListingCategory (numeric), Length: 84853, dtype: int64

As mentioned in the documnetation that the listing category column has some values i.e.
* ListingCategory 
The category of the listing that the borrower selected when posting their listing: 
* 0 - Not Available, 
* 1 - Debt Consolidation,
* 2 - Home Improvement, 
* 3 - Business, 
* 4 - Personal Loan,
* 5 - Student Use,
* 6 - Auto,
* 7- Other,
* 8 - Baby&Adoption, 
* 9 - Boat, 
* 10 - Cosmetic Procedure, 
* 11 - Engagement Ring, 
* 12 - Green Loans,
* 13 - Household E

In [37]:
# convert the ListingCategory (numeric) columns into the ListingCategory column with categories as metioned above 
categ = {0 : 'Not Available', 1 : 'Debt Consolidation', 2 : 'Home Improvement', 3: 'Business', 
             4 : 'Personal Loan', 5 : 'Student Use', 6 : 'Auto', 7 : 'Other', 8 : 'Baby&Adoption', 
             9 : 'Boat', 10 : 'Cosmetic Procedure', 11 : 'Engagement Ring', 12 : 'Green Loans',
             13 : 'Household Expenses', 14 : 'Large Purchases', 15 : 'Medical/Dental', 16 : 'Motorcycle',
             17 : 'RV', 18 : 'Taxes', 19 : 'Vacation', 20 : 'Wedding Loans'}

loan_df['ListingCategory'] = loan_df['ListingCategory (numeric)'].map(categ)
loan_df.drop(['ListingCategory (numeric)'], axis=1, inplace=True)

In [38]:
loan_df['ListingCategory'].unique()

array(['Home Improvement', 'Motorcycle', 'Debt Consolidation', 'Other',
       'Household Expenses', 'Auto', 'Medical/Dental', 'Wedding Loans',
       'Vacation', 'Taxes', 'Business', 'Baby&Adoption',
       'Engagement Ring', 'Large Purchases', 'Boat', 'RV',
       'Cosmetic Procedure', 'Not Available', 'Student Use',
       'Green Loans'], dtype=object)

In [39]:
# convert string to category type of ProsperRating (Alpha) column

values = ['AA', 'A', 'B', 'C', 'D', 'E','HR']
categorical_var = pd.api.types.CategoricalDtype(ordered = True, categories = values)
loan_df['ProsperRating'] = loan_df['ProsperRating (Alpha)'].astype(categorical_var)

In [40]:
loan_df.dtypes

ListingNumber                    object
ProsperRating (Alpha)            object
ProsperScore                      int32
Occupation                       object
EmploymentStatus                 object
IsBorrowerHomeowner                bool
CurrentCreditLines              float64
OpenRevolvingAccounts             int64
DebtToIncomeRatio               float64
IncomeVerifiable                   bool
LoanOriginalAmount                int64
LoanStatus                       object
CreditScoreRangeLower           float64
CreditScoreRangeUpper           float64
BorrowerState                    object
LoanOriginationDate      datetime64[ns]
IncomeRange                      object
BorrowerAPR                     float64
BorrowerRate                    float64
StatedMonthlyIncome             float64
Term                              int64
date                     datetime64[ns]
year                              int64
month                            object
ListingCategory                  object


In [41]:
# Converting the bool values into the numeric such that replacing the "True":1 and "False":0 of columns like ["IsBorrowerHomeowner","IncomeVerifiable"]IsBorrowerHomeowner","IncomeVerifiable"]

#df['col'] = df['col'].map({True: 1, False: 0})  this is the methof that I have used 

loan_df["IncomeVerifiable"]=loan_df["IncomeVerifiable"].map({True: 1, False:0})
# loan_df["IsBorrowerHomeowner"]=loan_df["IsBorrowerHomeowner"].map({True: 1, False:0})

In [42]:
loan_df["IsBorrowerHomeowner"]=loan_df["IsBorrowerHomeowner"].map({False: 0, True: 1})

In [43]:
loan_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 84853 entries, 1 to 113936
Data columns (total 26 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   ListingNumber          84853 non-null  object        
 1   ProsperRating (Alpha)  84853 non-null  object        
 2   ProsperScore           84853 non-null  int32         
 3   Occupation             84853 non-null  object        
 4   EmploymentStatus       84853 non-null  object        
 5   IsBorrowerHomeowner    84853 non-null  int64         
 6   CurrentCreditLines     84853 non-null  float64       
 7   OpenRevolvingAccounts  84853 non-null  int64         
 8   DebtToIncomeRatio      84853 non-null  float64       
 9   IncomeVerifiable       84853 non-null  int64         
 10  LoanOriginalAmount     84853 non-null  int64         
 11  LoanStatus             84853 non-null  object        
 12  CreditScoreRangeLower  84853 non-null  float64       
 13  

In [44]:
# calcute averge Credit Score Range
loan_df['CreditScore_average'] = (loan_df['CreditScoreRangeLower'] + loan_df['CreditScoreRangeUpper'])/2

In [45]:
loan_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 84853 entries, 1 to 113936
Data columns (total 27 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   ListingNumber          84853 non-null  object        
 1   ProsperRating (Alpha)  84853 non-null  object        
 2   ProsperScore           84853 non-null  int32         
 3   Occupation             84853 non-null  object        
 4   EmploymentStatus       84853 non-null  object        
 5   IsBorrowerHomeowner    84853 non-null  int64         
 6   CurrentCreditLines     84853 non-null  float64       
 7   OpenRevolvingAccounts  84853 non-null  int64         
 8   DebtToIncomeRatio      84853 non-null  float64       
 9   IncomeVerifiable       84853 non-null  int64         
 10  LoanOriginalAmount     84853 non-null  int64         
 11  LoanStatus             84853 non-null  object        
 12  CreditScoreRangeLower  84853 non-null  float64       
 13  

In [46]:
loan_df.describe()

ProsperScore  IsBorrowerHomeowner  CurrentCreditLines  \
count  84853.000000         84853.000000        84853.000000   
mean       5.950067             0.528538           10.508032   
std        2.376501             0.499188            5.317401   
min        1.000000             0.000000            0.000000   
25%        4.000000             0.000000            7.000000   
50%        6.000000             1.000000           10.000000   
75%        8.000000             1.000000           13.000000   
max       11.000000             1.000000           59.000000   

       OpenRevolvingAccounts  DebtToIncomeRatio  IncomeVerifiable  \
count           84853.000000       84853.000000      84853.000000   
mean                7.386292           0.258752          0.913580   
std                 4.515977           0.304333          0.280985   
min                 0.000000           0.000000          0.000000   
25%                 4.000000           0.160000          1.000000   
50%                 7.000000           0.240000          1.000000   
75%                10.000000           0.310000          1.000000   
max                50.000000          10.010000          1.000000   

       LoanOriginalAmount  CreditScoreRangeLower  CreditScoreRangeUpper  \
count        84853.000000           84853.000000           84853.000000   
mean          9083.440515             699.390240             718.390240   
std           6287.860058              47.095937              47.095937   
min           1000.000000             600.000000             619.000000   
25%           4000.000000             660.000000             679.000000   
50%           7500.000000             700.000000             719.000000   
75%          13500.000000             720.000000             739.000000   
max          35000.000000             880.000000             899.000000   

        BorrowerAPR  BorrowerRate  StatedMonthlyIncome          Term  \
count  84853.000000  84853.000000         8.485300e+04  84853.000000   
mean       0.226658      0.196022         5.931175e+03     42.486135   
std        0.079916      0.074631         8.239944e+03     11.640346   
min        0.045830      0.040000         0.000000e+00     12.000000   
25%        0.163280      0.135900         3.434000e+03     36.000000   
50%        0.219450      0.187500         5.000000e+03     36.000000   
75%        0.292540      0.257400         7.083333e+03     60.000000   
max        0.423950      0.360000         1.750003e+06     60.000000   

               year  CreditScore_average  
count  84853.000000         84853.000000  
mean    2012.328156           708.890240  
std        1.200926            47.095937  
min     2009.000000           609.500000  
25%     2012.000000           669.500000  
50%     2013.000000           709.500000  
75%     2013.000000           729.500000  
max     2014.000000           889.500000

In [48]:
# Create the new binary column 'LoanStatusBinary' based on your criteria
loan_df['LoanStatusBinary'] = loan_df['LoanStatus'].apply(lambda x: 1 if x == 'Completed' else 0)

# Drop the original 'LoanStatus' column
loan_df.drop(columns=['LoanStatus'], inplace=True)

# Rename the new column to 'LoanStatus'
loan_df.rename(columns={'LoanStatusBinary': 'LoanStatus'}, inplace=True)

In [49]:
loan_df

ListingNumber ProsperRating (Alpha)  ProsperScore  \
1            1209647                     A             7   
3             658116                     A             9   
4             909464                     D             4   
5            1074836                     B            10   
6             750899                     E             2   
...              ...                   ...           ...   
113932        753087                     C             5   
113933        537216                     A             8   
113934       1069178                     D             3   
113935        539056                     C             5   
113936       1140093                     A             7   

                     Occupation EmploymentStatus  IsBorrowerHomeowner  \
1                  Professional         Employed                    0   
3                 Skilled Labor         Employed                    1   
4                     Executive         Employed                    1   
5                  Professional         Employed                    1   
6                Sales - Retail         Employed                    0   
...                         ...              ...                  ...   
113932  Food Service Management         Employed                    1   
113933             Professional         Employed                    1   
113934                    Other         Employed                    1   
113935             Food Service        Full-time                    1   
113936                Professor         Employed                    0   

        CurrentCreditLines  OpenRevolvingAccounts  DebtToIncomeRatio  \
1                     14.0                     13               0.18   
3                      5.0                      7               0.15   
4                     19.0                      6               0.26   
5                     21.0                     13               0.36   
6                     10.0                      6               0.27   
...                    ...                    ...                ...   
113932                 9.0                      9               0.13   
113933                14.0                      9               0.11   
113934                12.0                      9               0.51   
113935                11.0                      7               0.48   
113936                10.0                      8               0.23   

        IncomeVerifiable  ...  BorrowerRate  StatedMonthlyIncome  Term  \
1                      1  ...        0.0920          6125.000000    36   
3                      1  ...        0.0974          2875.000000    36   
4                      1  ...        0.2085          9583.333333    36   
5                      1  ...        0.1314          8333.333333    60   
6                      1  ...        0.2712          2083.333333    36   
...                  ...  ...           ...                  ...   ...   
113932                 1  ...        0.1864          4333.333333    36   
113933                 1  ...        0.1110          8041.666667    36   
113934                 1  ...        0.2150          2875.000000    60   
113935                 1  ...        0.2605          3875.000000    60   
113936                 1  ...        0.1039          4583.333333    36   

             date  year month     ListingCategory  ProsperRating  \
1      2014-02-27  2014   Feb    Home Improvement              A   
3      2012-10-22  2012   Oct          Motorcycle              A   
4      2013-09-14  2013  Sept    Home Improvement              D   
5      2013-12-14  2013   Dec  Debt Consolidation              B   
6      2013-04-12  2013   Apr  Debt Consolidation              E   
...           ...   ...   ...                 ...            ...   
113932 2013-04-14  2013   Apr  Debt Consolidation              C   
113933 2011-11-03  2011   Nov               Other              A   
113934 2013-12-13  2013   Dec  Debt Consoli

In [51]:
# Find columns with missing values
columns_with_missing_values = loan_df.columns[loan_df.isnull().any()].tolist()

# Display the columns with missing values
print("Columns with missing values:", columns_with_missing_values)

Columns with missing values: []


In [52]:
# Identify columns with boolean values
boolean_columns = [col for col in loan_df.columns if loan_df[col].dtype == bool]

# Loop through boolean columns and cast as int
for col in boolean_columns:
    loan_df[col] = loan_df[col].astype(int)

In [54]:
loan_df.to_csv('preprocessed_data2.csv', index=False)